bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java

```diff
@@ -0,0 +1,411 @@
+    if (numPartitions == 0) {
+      // There are 0 partitions and 1 hashfile, return single source with full key range.
+      splitSources.add(
+          new KeyBasedHashTableSource(
+              conf,
+              hashTableOutputPathDir,
+              new ImmutableBytesWritable(hash.getStartRow()),
+              new ImmutableBytesWritable(hash.getStopRow())));
+      return splitSources;
+    }
+
+    // Use the HashTable start key. The value is HConstants.EMPTY_START_ROW for full table scan.
+    ImmutableBytesWritable startRow = new ImmutableBytesWritable(hash.getStartRow());
+    ImmutableBytesWritable stopRow = new ImmutableBytesWritable(hash.getStopRow());
+
+    // The output of HashTable is organized as partition file and a set of datafiles.
+    // Partition file contains a list of partitions, these partitions split the key-range of a table
+    // into roughly equal row-ranges and hashes for these row-ranges are stored in a single
+    // datafile.
+    //
+    // There are always numPartitions +1 data files. Datafile(i) covers hashes for [partition{i-1},
+    // partition{i}).
+    // So a partition file containing entries [b,f] for a table with row range [a,z] will have 3
+    // data files containing hashes.
+    // file0 will contain [a(startRow), b), file1 will contain [b,f), and file3 will contain
+    // [f,z(stopRow))
+    for (int i = 0; i < numPartitions; i++) {
+      LOG.debug(
+          "Adding: ["
+              + Bytes.toStringBinary(startRow.get())
+              + ", "
+              + Bytes.toStringBinary(partitions.get(i).get())
+              + "]");
+      splitSources.add(
+          new KeyBasedHashTableSource(conf, hashTableOutputPathDir, startRow, partitions.get(i)));
+      startRow = partitions.get(i);
+    }
+    // Add the last range for [lastPartition, stopRow).
+    LOG.debug(
+        "Adding: ["
+            + Bytes.toStringBinary(startRow.get())
+            + ", "
+            + Bytes.toStringBinary(stopRow.get())
+            + "]");
+    // Add the last range for [lastPartition, stopRow).
+    splitSources.add(
+        new KeyBasedHashTableSource(
+            conf,
+            hashTableOutputPathDir,
+            partitions.get(numPartitions - 1),
+            new ImmutableBytesWritable(stopRow)));
+    LOG.info("Returning " + splitSources.size() + " sources from " + numPartitions + " partitions");
+    return splitSources;
+  }
+
+  @Override
+  public Coder<RangeHash> getOutputCoder() {
+    return coder;
+  }
+
+  @Override
+  public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+    // HashTable data files don't expose a method to estimate size or lineCount.
+    return 0;
+  }
+
+  @Override
+  public BoundedReader<RangeHash> createReader(PipelineOptions options) {
+    // Reader should always come from KeyBasedSource.
+    throw new IllegalArgumentException("Reader can't be created from HadoopHashTableSource.");
```

Comment:
I used a similar format as CloudBigtableIO source, the dynamic parameters made it complicated. Can't do hash.getStartRow in constructor as hash is coming from a ValueProvider<> not available at the runtime. 

I merged the 2 classes, the constructors are not as clean as I would have liked, due to the valueproviders, but I think it will work.